### PR TITLE
Phase3

### PR DESCRIPTION
**Pull request by hu55a1n1**
Implements Phase 3 of #1 - Add support for IBC CreateClient and UpdateClient msgs.

- [x] complete/import the ics02/ics07 machinery from ibc-rs
- [x] state will now need to be able to store clients (may need more generic state mechanism ? )
- [x] allow sending MsgCreateClient to create client on the chain and MsgUpdateClient to update it with latest headers
- [x] should be able to use hermes tx raw create/update-client and/or gaiad tx ibc ...
- [x] should be able to run two chains, manually create a client on each, and manually update them

### CODE REVIEW

**Comment by hu55a1n1** [comment]
## Tested with the following setup

### Hermes config.toml
```toml
[global]
strategy = 'all'
log_level = 'trace'

[[chains]]
id = 'ibc-0'
rpc_addr = 'http://127.0.0.1:26657'
grpc_addr = 'http://127.0.0.1:9090'
websocket_addr = 'ws://localhost:26657/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas_price = { price = 0.001, denom = 'stake' }
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'

[[chains]]
id = 'ibc-1'
rpc_addr = 'http://127.0.0.1:26557'
grpc_addr = 'http://127.0.0.1:9091'
websocket_addr = 'ws://localhost:26557/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas_price = { price = 0.001, denom = 'stake' }
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'

[[chains]]
id = 'basecoin'
rpc_addr = 'http://127.0.0.1:26357'
grpc_addr = 'http://127.0.0.1:9093'
websocket_addr = 'ws://localhost:26357/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'basecoin'
#gas = 200000
#fee_denom = 'stake'
#fee_amount = 10
gas_price = { price = 0.001, denom = 'stake' }
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'
```

### Running basecoin
```sh
$ cargo run -- -v -p 26358 | grep 'basecoin::app' 
$ tendermint unsafe-reset-all; tendermint node --consensus.create_empty_blocks=false --rpc.laddr tcp://127.0.0.1:26357  # in a different shell
```

### Hermes commands
```sh
$ ./scripts/dev-env ~/.hermes/config.toml ibc-0 ibc-1
$ hermes tx raw create-client basecoin ibc-0
$ hermes tx raw update-client basecoin 07-tendermint-0
```

**Comment by thanethomson** [comment]
I'm assuming you need to modify your local `~/.tendermint/config/config.toml` to point to the basecoin-rs app at `tcp://127.0.0.1:26358` right?

Also, what `genesis.json` did you use locally for your basecoin-rs/Tendermint node? I tried supplying an empty (default) genesis file and basecoin-rs produced the following error:

```
Sep 20 15:20:03.711 DEBUG tendermint_abci::application: Incoming request: Request { value: Some(InitChain(RequestInitChain { time: Some(Timestamp { seconds: 1628284307, nanos: 287770036 }), chain_id: "test-chain-ssgPso", consensus_params: Some(ConsensusParams { block: Some(BlockParams { max_bytes: 22020096, max_gas: -1 }), evidence: Some(EvidenceParams { max_age_num_blocks: 100000, max_age_duration: Some(Duration { seconds: 172800, nanos: 0 }), max_bytes: 1048576 }), validator: Some(ValidatorParams { pub_key_types: ["ed25519"] }), version: None }), validators: [ValidatorUpdate { pub_key: Some(PublicKey { sum: Some(Ed25519([213, 18, 241, 105, 94, 83, 17, 229, 4, 7, 70, 72, 4, 26, 66, 164, 125, 146, 9, 125, 158, 45, 174, 38, 150, 52, 19, 248, 70, 234, 6, 137])) }), power: 10 }], app_state_bytes: [], initial_height: 1 })) }
Sep 20 15:20:03.711 DEBUG tendermint_basecoin::app: Got init chain request.
thread '<unnamed>' panicked at 'genesis state isn't valid JSON: Error("EOF while parsing a value", line: 1, column: 0)', src/app.rs:162:10
```

**Comment by hu55a1n1** [comment]
> I'm assuming you need to modify your local `~/.tendermint/config/config.toml` to point to the basecoin-rs app at `tcp://127.0.0.1:26358` right?

Right, I had to modify the following (also disabled Tendermint gRPC) ->
```toml
proxy_app = "tcp://127.0.0.1:26358"
laddr = "tcp://127.0.0.1:26357"
```
> Also, what genesis.json did you use locally for your basecoin-rs/Tendermint node? 

Here's my genesis.json `app_state` ->
```json
{
  "app_state": {
    "cosmos1snd5m4h0wt5ur55d47vpxla389r2xkf8dl6g9w": {
      "basecoin": 1000,
      "othercoin": 1000
    },
    "cosmos1t2e0nyjhwn3revunvf2uperhftvhzu4euuzva9": {
      "basecoin": 250,
      "othercoin": 5000
    },
    "cosmos1uawm90a5xm36kjmaazv89nxmfr8s8cyzkjqytd": {
      "acidcoin": 500
    },
    "cosmos1ny9epydqnr7ymqhmgfvlshp3485cuqlmt7vsmf": {},
    "cosmos1xwgdxu4ahd9eevtfnq5f7w4td3rqnph4llnngw": {
      "acidcoin": 500,
      "basecoin": 0,
      "othercoin": 100
    },
    "cosmos1mac8xqhun2c3y0njptdmmh3vy8nfjmtm6vua9u": {
      "basecoin": 1000
    },
    "cosmos1wkvwnez6fkjn63xaz7nzpm4zxcd9cetqmyh2y8": {
      "basecoin": 1
    },
    "cosmos166vcha998g7tl8j8cq0kwa8rfvm68cqmj88cff": {
      "basecoin": 18446744073709551615
    }
  }
}
```
I made an attempt to document this in the module specific READMEs -> [bank.md](https://github.com/informalsystems/basecoin-rs/blob/6f7bcecf69c0294c64028d5a04539852b57f9cea/docs/modules/bank.md) & [ibc.md](https://github.com/informalsystems/basecoin-rs/blob/6f7bcecf69c0294c64028d5a04539852b57f9cea/docs/modules/ibc.md).

File: src/app.rs [hunk 1]
```diff
@@ -1,133 +1,496 @@
...
+            "Got info request. Tendermint version: {}; Block version: {}; P2P version: {}, {:?}, {:?}",
+            request.version, request.block_version, request.p2p_version, last_block_height, last_block_app_hash
+        );
         ResponseInfo {
             data: "basecoin-rs".to_string(),
             version: "0.1.0".to_string(),
             app_version: 1,
             last_block_height,
-            last_block_app_hash,
+            last_block_app_hash: vec![],
```
Comment (thanethomson) [inline src/app.rs:154]: Any particular reason why this isn't set to the local `last_block_app_hash` variable value?

File: src/app.rs [hunk 2]
```diff
@@ -1,133 +1,496 @@
...
+                        code: 0,
+                        log: "exists".to_string(),
+                        info: "".to_string(),
+                        index: 0,
+                        key: request.data,
+                        value: result,
+                        proof_ops: None,
+                        height: self.store.read().unwrap().current_height() as i64,
+                        codespace: "".to_string(),
+                    };
```
Comment (thanethomson) [inline src/app.rs:203]:
```suggestion
                    return ResponseQuery {
                        code: 0,
                        log: "exists".to_string(),
                        key: request.data,
                        value: result,
                        height: self.store.read().unwrap().current_height() as i64,
                        ..ResponseQuery::default()
                    };
```

File: src/app.rs [hunk 3]
```diff
@@ -1,133 +1,496 @@
...
+                    app: 0,
+                }),
+                default_node_id: "".to_string(),
+                listen_addr: "".to_string(),
+                network: "".to_string(),
+                version: "".to_string(),
+                channels: vec![],
+                moniker: "".to_string(),
+                other: None,
+            }),
```
Comment (thanethomson) [inline src/app.rs:290]:
```suggestion
            default_node_info: Some(DefaultNodeInfo::default()),
```

File: src/app.rs [hunk 4]
```diff
@@ -1,133 +1,496 @@
...
+                git_commit: "209afef7e99ebcb814b25b6738d033aa5e1a932c".to_string(),
+                build_tags: "".to_string(),
+                go_version: "".to_string(),
+                build_deps: vec![VersionInfoModule {
+                    path: "github.com/cosmos/cosmos-sdk".to_string(),
+                    version: "v0.43.0".to_string(),
+                    sum: "h1:ps1QWfvaX6VLNcykA7wzfii/5IwBfYgTIik6NOVDq/c=".to_string(),
+                }],
+                cosmos_sdk_version: "".to_string(),
+            }),
```
Comment (thanethomson) [inline src/app.rs:304]:
```suggestion
            application_version: Some(VersionInfo {
                name: "basecoin-rs".to_string(),
                app_name: "basecoind".to_string(),
                version: "0.1.0".to_string(),
                git_commit: "209afef7e99ebcb814b25b6738d033aa5e1a932c".to_string(),
                build_deps: vec![VersionInfoModule {
                    path: "github.com/cosmos/cosmos-sdk".to_string(),
                    version: "v0.43.0".to_string(),
                    sum: "h1:ps1QWfvaX6VLNcykA7wzfii/5IwBfYgTIik6NOVDq/c=".to_string(),
                }],
                ..VersionInfo::default()
            }),
```

File: src/app.rs [hunk 5]
```diff
@@ -1,133 +1,496 @@
...
+        _request: Request<QueryAccountRequest>,
+    ) -> Result<Response<QueryAccountResponse>, Status> {
+        debug!("Got auth account request");
+
+        let account = BaseAccount {
+            address: "".to_string(),
+            pub_key: None,
+            account_number: 0,
+            sequence: 0,
+        };
```
Comment (thanethomson) [inline src/app.rs:364]:
```suggestion
        let account = BaseAccount::default();
```

File: src/app.rs [hunk 6]
```diff
@@ -1,133 +1,496 @@
...
+        Ok(Response::new(StakingQueryParamsResponse {
+            params: Some(Params {
+                unbonding_time: Some(Duration {
+                    seconds: 3 * 7 * 24 * 60 * 60,
+                    nanos: 0,
+                }),
+                max_validators: 0,
+                max_entries: 0,
+                historical_entries: 0,
+                bond_denom: "".to_string(),
```
Comment (thanethomson) [inline src/app.rs:492]:
```suggestion
                ..Params::default()
```

**Review by thanethomson — COMMENTED** [review]
So with the help of your instructions @hu55a1n1 I managed to get the various chains and the create/update client operations to work, thanks very much! :tada: 

I had a few (very) minor questions/suggestions here, but otherwise I think this is good to merge and move on to phase 4. We can always address any problems/improvements in follow-up PRs.

File: docs/modules/ibc.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+### Step 3: Configure hermes to be able to interact with basecoin
+```shell
+$ gaiad keys add user --keyring-backend="test" --output json > user_seed.json
+$ hermes keys add basecoin-0 -f user_seed.json
+```
+
+### Step 4: Create and Update a client
+```shell
+$ hermes tx raw create-client basecoin ibc-0
```
[the remark this answers is not in the document]
Comment (hu55a1n1) [reply docs/modules/ibc.md:50]: Thanks for reviewing @thanethomson! 💐 Just committed all your suggestions! 👌

File: src/app.rs (hunk 1 above)
Comment (hu55a1n1) [reply src/app.rs]: Nice catch! Done! 👍

**Pull request merged**